src/compiler/program.ts

```diff
@@ -2968,6 +2969,7 @@ namespace ts {
                 file.path = path;
                 file.resolvedPath = toPath(fileName);
                 file.originalFileName = originalFileName;
+                file.impliedNodeFormat = sourceFileOptions.impliedNodeFormat;

```

Comment:
Likewise, wasn't the file supposed to be (re)created with the correct format by the `getSourceFile` call above?

Reply:
This might not be handled in all calls and just like we ensuring say file paths, resolved paths etc its much easier to ensure that it is set.. (eg. in our repro - tsbuild uses sourceFile caching and it is irrespective of impliedNodeFormat .. the sharing of source file really just helps us with not having to reparse.. everything else including module resolution is redone any ways) so just setting the impliedNodeFormat here is better from perf standpoint

Reply:
We do this for eg for fileName, path etc properties  that we really care about being correct

Reply:
But that just means the program mutates the files when built, if two programs are actually getting used _at the same time_, this means one of them will have incorrect members and generate really really weird results, which is why it's critical that a separate source file be cached for each option value. For things like `filePath` it may not change much, maybe a path in an error message; but the `impliedNodeFormat` can change errors, emit, and how imports are handled - that's a _lot_ to suddenly get wrong when the language service (or any other multi-program environment) dumps a file into multiple programs.

Reply:
But that’s the thing .. in our scenario , ls creates new file but tsbuild is serialized operation and does not have two programs at the same time and helps with perf so I would like to keep same source file in this scenario ..

Reply:
This code is new to me, but Wesley's point is very compelling: it seems like this code is already trying to handle changes to `impliedNodeFormat`.  If that's not just a bug (I think Sheetal might be saying we might want to bypass the normal check for perf reasons?), it seems like the place that used to check the option ought to be the place that determines that a copy isn't necessary and the old node can just be updated.

Having said that, it looks like we're already clobbering a bunch of other properties on the reused node?  Maybe @weswigham is saying that those properties are less dangerous and this one needs special handling?

Reply:
Assuming this change makes sense, should we also set `languageVersion`?
```ts
newSourceFile.languageVersion = sourceFileOptions.languageVersion;
```

Reply:
Because we dont know what all might be impacted because of `impliedNodeFormat` differing i am reverting to my original fix of cache handling different `impliedNodeFormat`.